src/core/ctxt.ml

```diff
@@ -3,65 +3,59 @@
-let unbind ctx a b =
+let empty : ctxt = []
+
+(** [unbind ctx a def b] returns the triple [(x,b,ctx')] such that [(x,b)] is
+    the unbinding of [b] and [ctx'] is the context [ctx] extended with, if [x]
+    occurs in [b]
+    - the assumption that [x] is of type [a] if [def] is [None], and
+    - the definition of [x] in [def] of type [a] otherwise. *)
+let unbind : ctxt -> term -> term option -> tbinder -> tvar * term * ctxt =
+  fun ctx a def b ->
   let (x,b') = Bindlib.unbind b in
-  let ctx' = if Bindlib.binder_occur b then add x a ctx else ctx in
+  let ctx' = if Bindlib.binder_occur b then (x, a, def) :: ctx else ctx in
   (x,b',ctx')
 
-(** [pp oc ctx] prints the context [ctx] to the channel [oc]. *)
-let pp : t pp = fun oc ctx ->
-  let pp_e oc (x,a) =
-    Format.fprintf oc "%a : %a" Print.pp_tvar x Print.pp a
-  in
-  if ctx = [] then Format.pp_print_string oc "∅"
-  else List.pp pp_e ", " oc (List.rev ctx)
-
 (** [find x ctx] returns the type of [x] in the context [ctx] when it appears,
     and raises [Not_found] otherwise. *)
-let type_of : tvar -> t -> term = fun x ctx ->
-  snd (List.find (fun (y,_) -> Bindlib.eq_vars x y) ctx)
+let type_of : tvar -> ctxt -> term = fun x ctx ->
+  let (_,a,_) = List.find (fun (y,_,_) -> Bindlib.eq_vars x y) ctx in a
+
+(** [def_of x ctx] returns the definition of [x] in the context [ctx] if it
+    appears, and [None] otherwise *)
+let rec def_of : tvar -> ctxt -> term option = fun x ctx ->
+  match ctx with
+  | []         -> None
+  | (y,_,d)::l ->
+      if Bindlib.eq_vars x y then d else
+      def_of x l
 
 (** [mem x ctx] tells whether variable [x] is mapped in the context [ctx]. *)
-let mem : tvar -> t -> bool = fun x ctx ->
-  try ignore (type_of x ctx); false with Not_found -> true
+let mem : tvar -> ctxt -> bool = fun x ->
+  List.exists (fun (y,_,_) -> Bindlib.eq_vars x y)
 
 (** [to_prod ctx t] builds a product by abstracting over the context [ctx], in
     the term [t]. *)
-let to_prod : t -> term -> term = fun ctx t ->
-  match ctx with
-  | []      -> t
-  | [(x,a)] -> Prod(a, Bindlib.unbox (Bindlib.bind_var x (lift t)))
-  | _       -> let fn t (x,a) = _Prod (lift a) (Bindlib.bind_var x t) in
-               Bindlib.unbox (List.fold_left fn (lift t) ctx)
-
-(** [of_env] builds a context from an environment. **)
-let of_env : Env.t -> t =
-  List.map (fun (_,(v,bt)) -> v,Bindlib.unbox bt)
+let to_prod : ctxt -> term -> term = fun ctx t ->
+  let fn t elt =
+    match elt with
+    | (x,a,None   ) -> _Prod (lift a) (Bindlib.bind_var x t)
+    | (x,a,Some(u)) -> _LLet (lift u) (lift a) (Bindlib.bind_var x t)
+  in
+  Bindlib.unbox (List.fold_left fn (lift t) ctx)
 
 (** [make_meta ctx a] creates a metavariable of type [a],  with an environment
     containing the variables of context [ctx]. *)
-let make_meta : t -> term -> term = fun ctx a ->
+let make_meta : ctxt -> term -> term = fun ctx a ->
   let m = fresh_meta (to_prod ctx a) (List.length ctx) in
```

Comment:
Using List.length is incorrect now. to_prod should return both a term and the number of products. It's no problem since Ctxt.to_prod is only used in a few places in rewrite.ml and handle.ml to build terms.

In fact, now that ctxt is defined in terms.ml and ctxt.ml is a small file, I suggest to rename to_prod into prod, move all the contents of ctxt.ml into basics.ml, and remove ctxt.ml.

Reply:
Some simplifications can be done in basics.ml as well:
- remove iter_ctxt which is unused
- move appl_to_tref to eval.ml